File: Tree/Red-BlackTree/Red-BlackTree.cpp
#include <iostream>

enum Color { RED, BLACK };

class Node {
public:
    int key;
    Color color;
    Node* left;
    Node* right;
    Node* parent;

    Node(int k, Color c = RED) : key(k), color(c), left(nullptr), right(nullptr), parent(nullptr) {}
};

class RedBlackTree {
private:
    Node* root;
    Node* NIL;

    void leftRotate(Node* x) {
        Node* y = x->right;
        x->right = y->left;

        if (y->left != NIL)
            y->left->parent = x;

        y->parent = x->parent;

        if (x->parent == nullptr)
            root = y;
        else if (x == x->parent->left)
            x->parent->left = y;
        else
            x->parent->right = y;

        y->left = x;
        x->parent = y;
    }

    void rightRotate(Node* y) {
        Node* x = y->left;
        y->left = x->right;

        if (x->right != NIL)
            x->right->parent = y;

        x->parent = y->parent;

        if (y->parent == nullptr)
            root = x;
        else if (y == y->parent->left)
            y->parent->left = x;
        else
            y->parent->right = x;

        x->right = y;
        y->parent = x;
    }

    void insertFixup(Node* z) {
        while (z->parent->color == RED) {
            if (z->parent == z->parent->parent->left) {
                Node* y = z->parent->parent->right;
                if (y->color == RED) {
                    z->parent->color = BLACK;
                    y->color = BLACK;
                    z->parent->parent->color = RED;
                    z = z->parent->parent;
                } else {
                    if (z == z->parent->right) {
                        z = z->parent;
                        leftRotate(z);
                    }

                    z->parent->color = BLACK;
                    z->parent->parent->color = RED;
                    rightRotate(z->parent->parent);
                }
            } else {
                Node* y = z->parent->parent->left;
                if (y->color == RED) {
                    z->parent->color = BLACK;
                    y->color = BLACK;
                    z->parent->parent->color = RED;
                    z = z->parent->parent;
                } else {
                    if (z == z->parent->left) {
                        z = z->parent;
                        rightRotate(z);
                    }

                    z->parent->color = BLACK;
                    z->parent->parent->color = RED;
                    leftRotate(z->parent->parent);
                }
            }
        }

        root->color = BLACK;
    }

    void insert(Node* z) {
        Node* y = NIL;
        Node* x = root;

        while (x != NIL) {
            y = x;
            if (z->key < x->key)
                x = x->left;
            else
                x = x->right;
        }

        z->parent = y;
        if (y == NIL)
            root = z;
        else if (z->key < y->key)
            y->left = z;
        else
            y->right = z;

        z->left = NIL;
        z->right = NIL;
        z->color = RED;

        insertFixup(z);
    }

    void inorderTraversal(Node* node) {
        if (node != NIL) {
            inorderTraversal(node->left);
            std::cout << node->key << " ";
            inorderTraversal(node->right);
        }
    }

public:
    RedBlackTree() {
        NIL = new Node(-1, BLACK);
        root = NIL;
    }

    void insert(int key) {
        Node* z = new Node(key);
        insert(z);
    }

    void printInOrderTraversal() {
        inorderTraversal(root);
    }
};

int main() {
    RedBlackTree rbTree;
    rbTree.insert(10);
    rbTree.insert(5);
    rbTree.insert(15);
    rbTree.insert(3);
    rbTree.insert(7);

    std::cout << "Inorder traversal: ";
    rbTree.printInOrderTraversal();
    std::cout << std::endl;

    return 0;
}

File: Tree/Red-BlackTree/Red-BlackTree.cs
using System;

enum Color
{
    Red,
    Black
}

class RedBlackNode
{
    public int Key { get; set; }
    public Color NodeColor { get; set; }
    public RedBlackNode Parent { get; set; }
    public RedBlackNode Left { get; set; }
    public RedBlackNode Right { get; set; }

    public RedBlackNode(int key, Color color)
    {
        Key = key;
        NodeColor = color;
        Parent = null;
        Left = null;
        Right = null;
    }
}

class RedBlackTree
{
    private RedBlackNode root;

    private void LeftRotate(RedBlackNode x)
    {
        RedBlackNode y = x.Right;
        x.Right = y.Left;

        if (y.Left != null)
            y.Left.Parent = x;

        y.Parent = x.Parent;

        if (x.Parent == null)
            root = y;
        else if (x == x.Parent.Left)
            x.Parent.Left = y;
        else
            x.Parent.Right = y;

        y.Left = x;
        x.Parent = y;
    }

    private void RightRotate(RedBlackNode y)
    {
        RedBlackNode x = y.Left;
        y.Left = x.Right;

        if (x.Right != null)
            x.Right.Parent = y;

        x.Parent = y.Parent;

        if (y.Parent == null)
            root = x;
        else if (y == y.Parent.Left)
            y.Parent.Left = x;
        else
            y.Parent.Right = x;

        x.Right = y;
        y.Parent = x;
    }

    private void InsertFixup(RedBlackNode z)
    {
        while (z.Parent != null && z.Parent.NodeColor == Color.Red)
        {
            if (z.Parent == z.Parent.Parent.Left)
            {
                RedBlackNode y = z.Parent.Parent.Right;

                if (y != null && y.NodeColor == Color.Red)
                {
                    z.Parent.NodeColor = Color.Black;
                    y.NodeColor = Color.Black;
                    z.Parent.Parent.NodeColor = Color.Red;
                    z = z.Parent.Parent;
                }
                else
                {
                    if (z == z.Parent.Right)
                    {
                        z = z.Parent;
                        LeftRotate(z);
                    }

                    z.Parent.NodeColor = Color.Black;
                    z.Parent.Parent.NodeColor = Color.Red;
                    RightRotate(z.Parent.Parent);
                }
            }
            else
            {
                RedBlackNode y = z.Parent.Parent.Left;

                if (y != null && y.NodeColor == Color.Red)
                {
                    z.Parent.NodeColor = Color.Black;
                    y.NodeColor = Color.Black;
                    z.Parent.Parent.NodeColor = Color.Red;
                    z = z.Parent.Parent;
                }
                else
                {
                    if (z == z.Parent.Left)
                    {
                        z = z.Parent;
                        RightRotate(z);
                    }

                    z.Parent.NodeColor = Color.Black;
                    z.Parent.Parent.NodeColor = Color.Red;
                    LeftRotate(z.Parent.Parent);
                }
            }
        }

        root.NodeColor = Color.Black;
    }

    public void Insert(int key)
    {
        RedBlackNode z = new RedBlackNode(key, Color.Red);
        RedBlackNode y = null;
        RedBlackNode x = root;

        while (x != null)
        {
            y = x;
            if (z.Key < x.Key)
                x = x.Left;
            else
                x = x.Right;
        }

        z.Parent = y;
        if (y == null)
            root = z;
        else if (z.Key < y.Key)
            y.Left = z;
        else
            y.Right = z;

        InsertFixup(z);
    }

    private void InOrderTraversal(RedBlackNode node)
    {
        if (node != null)
        {
            InOrderTraversal(node.Left);
            Console.WriteLine($"{node.Key} ({node.NodeColor})");
            InOrderTraversal(node.Right);
        }
    }

    public void PrintInOrderTraversal()
    {
        InOrderTraversal(root);
    }
}

class Program
{
    static void Main()
    {
        RedBlackTree rbTree = new RedBlackTree();
        rbTree.Insert(10);
        rbTree.Insert(5);
        rbTree.Insert(15);
        rbTree.Insert(3);
        rbTree.Insert(7);

        Console.WriteLine("Inorder traversal:");
        rbTree.PrintInOrderTraversal();
    }
}